Mathlib/CategoryTheory/MorphismProperty/Lifting.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Limits
+import Mathlib.CategoryTheory.MorphismProperty.Retract
+
+/-!
+# Left and right lifting properties
+
+Given a morphism property `T`, we define the left and right lifting property with respect to `T`.
+
+We show that the left lifting property is stable under retracts, cobase change, and composition,
+with dual statements for the right lifting property.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C] (T : MorphismProperty C)
+
+namespace MorphismProperty
+
+/-- The left lifting property (llp) with respect to a class of morphisms. -/
+def llp : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ (g : X ⟶ Y) (_ : T g), HasLiftingProperty f g
+
+/-- The right lifting property (rlp) with respect to a class of morphisms. -/
+def rlp : MorphismProperty C := fun _ _ f ↦
+  ∀ ⦃X Y : C⦄ (g : X ⟶ Y) (_ : T g), HasLiftingProperty g f
+
+lemma _root_.CategoryTheory.RetractArrow.leftLiftingProperty
+    {X Y Z W Z' W' : C} {g : Z ⟶ W} {g' : Z' ⟶ W'}
+    (h : RetractArrow g' g) (f : X ⟶ Y) [HasLiftingProperty g f] : HasLiftingProperty g' f where
+  sq_hasLift := fun {u v} sq ↦ by
+    have sq' : CommSq (h.r.left ≫ u) g f (h.r.right ≫ v) := by simp [sq.w]
+    exact
+      ⟨⟨{ l := h.i.right ≫ sq'.lift
+          fac_left := by
+            simp only [← h.i_w_assoc, sq'.fac_left, h.retract_left_assoc,
+              Arrow.mk_left, Category.id_comp]
+          fac_right := by simp}⟩⟩
+
+lemma _root_.CategoryTheory.RetractArrow.rightLiftingProperty
+    {X Y Z W X' Y' : C} {f : X ⟶ Y} {f' : X' ⟶ Y'}
+    (h : RetractArrow f' f) (g : Z ⟶ W) [HasLiftingProperty g f] : HasLiftingProperty g f' where
+  sq_hasLift := fun {u v} sq ↦ by
+    have sq' : CommSq (u ≫ h.i.left) g f (v ≫ h.i.right) :=
+      ⟨by rw [← Category.assoc, ← sq.w, Category.assoc, RetractArrow.i_w, Category.assoc]⟩
+    exact
+      ⟨⟨{ l := sq'.lift ≫ h.r.left
+          fac_left := by simp
```

Comment:
In the definition of `LiftStruct` in the file `CommSq.lean`, we should add `:= by aesop_cat` as the default value of `fac_left` and `fac_right`, so that some proofs could be found automatically.